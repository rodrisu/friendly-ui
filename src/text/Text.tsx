import React from 'react'
import cc from 'classcat'

import { replaceNewLineWithBR } from '../_utils'
import { A11yProps, pickA11yProps } from '../_utils/interfaces'

export enum TextDisplayType {
  BODY = 'body',
  BODYSTRONG = 'bodyStrong',
  BUTTON = 'button',
  CAPTION = 'caption',
  DISPLAY1 = 'display1',
  DISPLAY2 = 'display2',
  SUBHEADER = 'subheader',
  SUBHEADERSTRONG = 'subheaderStrong',
  TITLE = 'title',
  TITLESTRONG = 'titleStrong',
}

export enum TextTagType {
  DIV = 'div',
  PARAGRAPH = 'p',
  SPAN = 'span',
}

export type TextProps = A11yProps &
  Readonly<{
    className?: string
    children: string | number | React.ReactNode
    display?: TextDisplayType
    tag?: TextTagType
    textColor?: string
    newlineToBr?: boolean
    itemProp?: string
  }>

const baseClassName = 'kirk-text'
const cssColorRegex = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i

export const Text = (props: TextProps) => {
  const {
    className,
    children,
    display = TextDisplayType.BODY,
    tag = TextTagType.SPAN,
    textColor,
    newlineToBr = true,
    itemProp = null,
  } = props
  const a11yAttrs = pickA11yProps<TextProps>(props)
  const displayClassName = `${baseClassName}-${display}`
  const Tag = tag

  const inlineStyle = cssColorRegex.test(textColor) ? { style: { color: textColor } } : null

  return (
    <Tag
      className={cc([baseClassName, displayClassName, className])}
      {...inlineStyle}
      {...(itemProp && { itemProp })}
      {...a11yAttrs}
    >
      {typeof children === 'string' && newlineToBr ? replaceNewLineWithBR(children) : children}
    </Tag>
  )
}
