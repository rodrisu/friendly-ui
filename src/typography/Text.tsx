import React, { ReactNode } from 'react'
import { ThemeProps } from 'styled-components'

import { replaceNewLineWithBR } from '../_utils'

export type TextProps = Readonly<{
  id?: string
  className?: string
  children: ReactNode
  itemProp?: string
  theme: ThemeProps<any>
}>

export const Text = ({ theme, ...props }: TextProps): JSX.Element => {
  const children =
    typeof props.children === 'string' ? replaceNewLineWithBR(props.children) : props.children
  return React.cloneElement(<span />, props, children)
}
