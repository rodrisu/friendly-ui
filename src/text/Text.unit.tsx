import * as React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { Text, TextTagType } from './Text'

const multipleLineText = `line1
line2
line3`

describe('Text', () => {
  it('should have change the DOM', () => {
    const text = renderer.create(<Text>text</Text>).toJSON()
    expect(text).toMatchSnapshot()
  })

  it('Should have the custom class', () => {
    const wrapper = shallow(<Text className="testClass">test</Text>)
    expect(wrapper.hasClass('testClass')).toBe(true)
  })

  it('Should render with the default span tag', () => {
    const wrapper = shallow(<Text>test</Text>)
    expect(wrapper.name()).toBe('span')
  })

  it('Should render with the div tag', () => {
    const wrapper = shallow(<Text tag={TextTagType.DIV}>test</Text>)
    expect(wrapper.name()).toBe('div')
  })

  it('Should have role attribute when present and undefined otherwise', () => {
    const wrapper = shallow(<Text>test</Text>)
    expect(wrapper.prop('role')).toBeUndefined()

    wrapper.setProps({ role: 'alert' })

    expect(wrapper.prop('role')).toBe('alert')
  })

  it('Should have aria-label attribute when present and undefined otherwise', () => {
    const wrapper = shallow(<Text>test</Text>)
    expect(wrapper.prop('aria-label')).toBeUndefined()

    wrapper.setProps({ 'aria-label': 'Complementary description' })

    expect(wrapper.prop('aria-label')).toBe('Complementary description')
  })

  it('Should accept ReactNode as children', () => {
    const wrapper = shallow(
      <Text tag={TextTagType.DIV}>
        <span>test</span>
      </Text>,
    )
    expect(wrapper.children().first().name()).toBe('span')
  })

  it('Should replace \\n with br tags by default', () => {
    const wrapper = shallow(<Text>{multipleLineText}</Text>)
    expect(wrapper.html()).toContain(
      '<span class="kirk-text kirk-text-body">line1<br/>line2<br/>line3</span>',
    )
  })

  it('Should not replace \\n with br tags when newlineToBr is false', () => {
    const wrapper = shallow(<Text newlineToBr={false}>{multipleLineText}</Text>)
    expect(wrapper.html()).toContain(
      `<span class="kirk-text kirk-text-body">${multipleLineText}</span>`,
    )
  })

  it('Should display the text with color', () => {
    const wrapper = shallow(<Text textColor="#FFFFFF">test</Text>)
    expect(wrapper.prop('style')).toEqual({ color: '#FFFFFF' })
  })

  it('Should not display the text with color when color not hexa', () => {
    const wrapper = shallow(<Text textColor="#FFFGFF">test</Text>)
    expect(wrapper.prop('style')).toBeUndefined()
  })
})
