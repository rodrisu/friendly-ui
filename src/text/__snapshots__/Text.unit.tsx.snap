// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should have change the DOM 1`] = `
<span
  className="kirk-text kirk-text-body"
>
  text
</span>
`;
